docker.md

```diff
@@ -1,184 +0,0 @@
-
-Containers are based on **images** which are immutable filesystem layers that define an application's environment. The application itself is added on top of an existing base image like `ubuntu` or `centos` for instance. These are publicly available.
-
-The running state of our system is two docker containers, one running Gem (just telnetd for now) and one running Postgres. The Postgres container uses the `postgres:9.6.0` image which is maintained by Postgres. The Gem container uses the `java:latest` base image with our application added on via sbt-native-packager, yielding (for now) an image called `telnetd:0.1-SNAPSHOT`.
-
-### Getting Started
-
-Ok what we want to do is `./gemctl deploy` to make our first deployment. So try that.
-
-    gem$ ./gemctl deploy
-    [gemctl] ERROR: Docker isn't installed or isn't running. Download at https://www.docker.com/
-    gem$
-
-Ok so go install Docker, start it, and try again.
-
-    gem$ ./gemctl deploy
-    [gemctl] Creating gem-net bridge network.
-    [gemctl] ERROR: Could not find telnetd:0.1-SNAPSHOT image. You may need to do 'sbt docker:publishLocal'.
-    gem$
-
-We got a little farther this time. Looks like we need to publish our image. This will build the app and add it to our local Docker image repository. So do what it says and try again.
-
-    gem$ ./gemctl deploy
-    gem (docker *)$ ./gemctl deploy
-    [gemctl] Found gem-net bridge network.
-    [gemctl] Found telnetd:0.1-SNAPSHOT image.
-    [gemctl] This is the first deployment on localhost.
-    [gemctl] Deploying Postgres container db-1.
-    [gemctl] Creating database.
-    [gemctl] Creating Gem container gem-1.
-    [gemctl] Telnet is exposed on 1234.
-    [gemctl] TODO: Import existing data. We will need a mount point and machinery for this.
-    [gemctl] Verifying deployment 1.
-    [gemctl] Deployment 1 database looks ok.
-    [gemctl] Deployment 1 Gem server looks ok.
-    gem$
-
-And test it:
-
-    gem$ telnet localhost 1234
-    Trying ::1...
-    Connected to localhost.
-    Escape character is '^]'.
-    Welcome to Gem
-    Username: root
-    Password:
-    Welcome, local time is Tue Dec 13 21:58:36 UTC 2016.
-    root@gem> exit
-    Goodbye.
-    Connection closed by foreign host.
-    gem$
-
-We can look at the running containers.
-
-    gem$ docker ps
-    CONTAINER ID        IMAGE                  COMMAND                  CREATED              STATUS              PORTS                    NAMES
-    e4b079b1c2a7        telnetd:0.1-SNAPSHOT   "bin/telnetd"            About a minute ago   Up About a minute   0.0.0.0:1234->6666/tcp   gem-1
-    2231cdc6ef9e        postgres:9.6.0         "/docker-entrypoint.s"   About a minute ago   Up About a minute   5432/tcp                 db-1
```

Comment:
No, the port is local to the container. All the containers are on a private virtual network called `gem-net` and they can see each other and refer to each other by name. So the `gem-X` container refers to Postgres host `db-X` as if it were another machine.

This is how upgrade deployments work. We start up a new database server `db-42` and then stream all the data from `db-41` (or whatever the previous one is called) before shutting it down.

If we want to expose ports to the outside world on non-random ports then yes you can have a conflict, as it the case of the public telnet service mapped to localhost:6666. If you try to start up another gem container it will fail because the public port is taken.